### DISCUSSION TITLE
能否利用Xhttp上下行实现A/B两个服务器穿透组网？

### DISCUSSION THREAD

**Discussion by Finer3004** [XTLS/Xray-core#4116] [Q&A]
假设我有A和B两个机器，A有公网，并且可以用过内网连接到B，B公网inbound被block，但是公网outbound可用。可否利用xhttp实现：客户端走reality入，上行传输到A；然后A传输到B，出的话，下行通过B的outbound传输到客户端。我希望利用B的流量和线路来下行。请问我的设想可否实现？并且需要怎么改动。

我发现之前有一个解决方案是走CDN入，直连出，但是处于同一个机器： https://github.com/XTLS/Xray-core/discussions/4084 。

**Comment by lxhao61** [comment]


**Reply by Finer3004** [reply]
看到一个方案：https://blog.njj0121.com/proxy/2024/11/07/xray-xhttpreality%e4%b8%8a%e4%b8%8b%e8%a1%8c%e5%88%86%e7%a6%bb/1805/ 
貌似更简单？：A直接充当任意门外网转发到内网的B，B正常reality偷一个证书（没有Inbound搞不下证书来自己偷自己），通过path再回落到xhttp，客户端对接B的reality和xhttp配置。

**Reply by Finer3004** [reply]
有一个共有的问题，参看：https://github.com/XTLS/Xray-core/discussions/4113 。RPRX说为了分离需要在客户端的 "downloadSettings": 设置"address": "", // another domain/IP“和"port"。但是B的inboud端口被封了，无法获得B的port。是不是无法实现从B下行？最终还是从A下行。

**Reply by lxhao61** [reply]
晕，我之前理解错了。是你 B 被墙了，只能 B 套 CDN 才可实现你想要的上下行分离，或 A 中转流量给 B 来用（不能上下行分离）。

**Comment by RPRX** [comment]
> B公网inbound被block，但是公网outbound可用

这相当于内网，你需要的是 UDP 打洞

**Reply by lxhao61** [reply]
@RPRX 大神，见  https://github.com/XTLS/Xray-core/commit/ae62a0fb526f2faa0aa8b3fbce5a566187059090#comments 留言，什么原因无法兼容？

**Reply by RPRX** [reply]
@lxhao61 我已回复，在那里继续讨论

**Reply by lxhao61** [reply]
> @lxhao61 我已回复，在那里继续讨论

不知道怎么回事那里我看不了，整个无法显示出来了。
![Screenshot_2024-12-06-01-53-35-592_org mozilla fi](https://github.com/user-attachments/assets/c64e6d56-818a-4950-a531-884bb07e7381)

**Reply by RPRX** [reply]
https://github.com/XTLS/Xray-core/commit/ae62a0fb526f2faa0aa8b3fbce5a566187059090#commitcomment-150004845

> 但是 XHTTP 会把 `/yourpath` 解析成 `/yourpath/`，所以 HTTP 的 path 需要以 `/` 结尾

**Reply by lxhao61** [reply]
> https://github.com/XTLS/Xray-core/commit/ae62a0fb526f2faa0aa8b3fbce5a566187059090#commitcomment-150004845
> 
> > 但是 XHTTP 会把 `/yourpath` 解析成 `/yourpath/`，所以 HTTP 的 path 需要以 `/` 结尾

原来这样，己测试通过。

**Comment by Finer3004** [comment]
请问Xray可以实现UDP打洞吗？有相关配置介绍吗。

**Comment by Smallthing** [comment]
看了你这个强大的服务器网络拓扑,感觉是灰色以上的产业,哈哈哈
